Kernel/Devices/USB/UHCIController.cpp

```diff
@@ -0,0 +1,470 @@
+    setup_schedule();
+
+    write_flbaseadd(m_framelist->physical_page(0)->paddr().get()); // Frame list (physical) address
+    write_frnum(0);                                                // Set the initial frame number
+
+    // Enable all interrupt types
+    write_frnum(UHCI_USBINTR_TIMEOUT_CRC_ENABLE | UHCI_USBINTR_RESUME_INTR_ENABLE | UHCI_USBINTR_IOC_ENABLE | UHCI_USBINTR_SHORT_PACKET_INTR_ENABLE);
+    klog() << "UHCI: Reset completed!";
+}
+
+void UHCIController::create_structures()
+{
+    // Let's allocate memory for botht the QH and TD pools
+    // First the QH pool and all of the Interrupt QH's
+    auto qh_pool_vmobject = ContiguousVMObject::create_with_size(2 * PAGE_SIZE);
+    m_qh_pool = MemoryManager::the().allocate_kernel_region_with_vmobject(*qh_pool_vmobject, 2 * PAGE_SIZE, "UHCI Queue Head Pool", Region::Access::Write);
+    memset(m_qh_pool->vaddr().as_ptr(), 0, 2 * PAGE_SIZE); // Zero out both pages
+
+    // Let's populate our free qh list (so we have some we can allocate later on)
+    m_free_qh_pool.resize(MAXIMUM_NUMBER_OF_TDS);
+    for (size_t i = 0; i < m_free_qh_pool.size(); i++) {
+        auto placement_addr = reinterpret_cast<void*>(m_qh_pool->vaddr().get() + (i * sizeof(QueueHead)));
+        auto paddr = static_cast<u32>(m_qh_pool->physical_page(0)->paddr().get() + (i * sizeof(QueueHead)));
+        m_free_qh_pool.at(i) = new (placement_addr) QueueHead(paddr);
+
+        //auto& queue_head = m_free_qh_pool.at(i);
+    }
+
+    // Create the Interrupt Transfer, Full Speed/Low Speed Control and Bulk Queue Heads
+    m_interrupt_transfer_queue = allocate_queue_head();
+    m_lowspeed_control_qh = allocate_queue_head();
+    m_fullspeed_control_qh = allocate_queue_head();
+    m_bulk_qh = allocate_queue_head();
+    m_dummy_qh = allocate_queue_head();
+
+    // Now the Transfer Descriptor pool
+    auto td_pool_vmobject = ContiguousVMObject::create_with_size(2 * PAGE_SIZE);
+    m_td_pool = MemoryManager::the().allocate_kernel_region_with_vmobject(*td_pool_vmobject, 2 * PAGE_SIZE, "UHCI Transfer Descriptor Pool", Region::Access::Write);
+    memset(m_td_pool->vaddr().as_ptr(), 0, 2 * PAGE_SIZE);
+
+    // Set up the Isochronous Transfer Descriptor list
+    m_iso_td_list.resize(UHCI_NUMBER_OF_ISOCHRONOUS_TDS);
+    for (size_t i = 0; i < m_iso_td_list.size(); i++) {
+        auto placement_addr = reinterpret_cast<void*>(m_td_pool->vaddr().get() + (i * sizeof(Kernel::USB::TransferDescriptor)));
+        auto paddr = static_cast<u32>(m_td_pool->physical_page(0)->paddr().get() + (i * sizeof(Kernel::USB::TransferDescriptor)));
+
+        // Place a new Transfer Descriptor with a 1:1 in our region
+        // The pointer returned by `new()` lines up exactly with the value
+        // that we store in `paddr`, meaning our member functions directly
+        // access the raw descriptor (that we later send to the controller)
+        m_iso_td_list.at(i) = new (placement_addr) Kernel::USB::TransferDescriptor(paddr);
```

Comment:
I feel like this is a neat trick to map these structures virtual->physical, which makes life a lot easier.